test/jdk/javax/swing/JTable/JTableRightAlignmentTest.java

```diff
@@ -0,0 +1,200 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.awt.Color;
+import java.awt.ComponentOrientation;
+import java.awt.GridBagConstraints;
+import java.awt.GridBagLayout;
+import java.awt.Insets;
+import java.awt.Rectangle;
+import java.awt.Robot;
+import java.awt.image.BufferedImage;
+
+import javax.swing.JFrame;
+import javax.swing.JPanel;
+import javax.swing.JScrollPane;
+import javax.swing.JTable;
+import javax.swing.SwingUtilities;
+import javax.swing.table.AbstractTableModel;
+import javax.swing.UIManager;
+
+import javax.imageio.ImageIO;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/*
+ * @test
+ * @key headful
+ * @bug 5108458
+ * @summary Test to check Right alignment of JTable data
+ * (Fix affects all L&F, test verifies only Metal L&F)
+ * @run main JTableRightAlignmentTest
+ */
+
+public class JTableRightAlignmentTest {
+    static JFrame frame;
+    static CustomTable table;
+    static boolean passed = true;
+    static String failureString;
+
+    public static void main(String[] args) throws Exception {
+        UIManager.setLookAndFeel("javax.swing.plaf.metal.MetalLookAndFeel");
+        Robot robot = new Robot();
+        try {
+            SwingUtilities.invokeAndWait(() -> {
+                frame = new JFrame("Test JTable");
+
+                JPanel panel = new JPanel(new GridBagLayout());
+                frame.setContentPane(panel);
+                table = new CustomTable();
+                panel.add(new JScrollPane(table.table),
+                        new GridBagConstraints(0, 0, -1, -1, 1.0, 1.0,
+                                GridBagConstraints.PAGE_START, GridBagConstraints.BOTH,
+                                new Insets(2, 2, 2, 2), 0, 0));
+                frame.pack();
+                frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
+                frame.applyComponentOrientation(ComponentOrientation.RIGHT_TO_LEFT);
+                frame.setVisible(true);
+            });
+
+            robot.waitForIdle();
+            robot.delay(1000);
+            SwingUtilities.invokeAndWait(() -> {
+                int maxHeight = (int) (((double) table.table.getTableHeader().getHeight())
+                        + ((double) table.table.getHeight()));
+                int yPos = table.table.getTableHeader().getLocationOnScreen().y;
+                int xPos = table.table.getLocationOnScreen().x + table.table.getWidth() -
+                        table.table.getTableHeader().getColumnModel()
+                                .getColumn(2)
+                                .getWidth() - 1;
+                Color expectedRGB = robot.getPixelColor(xPos, yPos);
```

Comment:
I suggest capturing the screen rectangle and then sample the colours of the captured buffered image, or alternatively even paint into a buffered image.

Reply:
Any particular reason for this suggestion ?

Reply:
It is much more efficient this way: `getPixelColor` captures 1 pixel of the screen but what you really want is to capture the entire table one way or another. Capturing the entire table with `createScreenCapture` once means you read pixels from the screen only once.

Painting to a `BufferedImage` is even more efficient, it avoids accessing screen pixels altogether, and the test could be made headless.